reps/2023-03-15-train-api.md

```diff
@@ -0,0 +1,162 @@
+# Simplifying Ray Train Ingest APIs
+
+## Summary
+Deprecate the preprocessor and object store memory arguments for `ray.train.Trainer`.
+
+### General Motivation
+
+This doc proposes to simplify Train's DatasetConfig as we move to the new streaming backend by default for Datasets. Similar to as noted in https://github.com/ray-project/enhancements/pull/25, Ray Datasets will have both lazy and streaming execution by default in Ray 2.4. Furthermore, `DatasetPipeline` will be deprecated in the future to consolidate functionality on the Dataset class.
+
+With these changes, a few possibilities for simplification open up in the Train API:
+1. Decoupling Preprocessors from Trainers, so that Data preprocessing is performed on the Dataset explicitly by the user prior to passing the Dataset into the Trainer.
+2. Using the same resource limiting API in Train as in Datasets (i.e., `ExecutionResources`), instead of having a separate `max_object_store_memory_fraction` config.
+
+Simplification is greatly desirable here, since users commonly find Dataset\<\>Trainer interactions difficult to understand and debug.
+
+### Should this change be within `ray` or outside?
+main `ray` project. Changes are made to Ray Data and Ray AIR level.
+
+## Stewardship
+### Required Reviewers
+The proposal will be open to the public, but please suggest a few experienced Ray contributors in this technical domain whose comments will help this proposal. Ideally, the list should include Ray committers.
+
+@amogkam, @c21, @clarkzinzow, @jianoiax
+
+### Shepherd of the Proposal (should be a senior committer)
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@stephanie-wang
+
+## Design and Architecture
+
+## API Changes
+
+1. Introduce a `resource_limits: ExecutionResources(object_store_memory=2 * GiB)` arg to `ray.air.DatasetConfig`. This enables streaming by default, with a limit of 2GiB, and deprecates the previous `max_object_store_memory_fraction` argument.
+
+2. Introduce `Dataset.get_logical_plan()` (DeveloperAPI), which returns the logical plan that can be used to extract the lineage of preprocessors applied to this Dataset. If multiple preprocessors are applied, Train can return a `Chain` of the preprocessors. Non-preprocessor operations on the Dataset are ignored, and we can also allow ignoring preprocessors such as per epoch preprocessors. This function will be used by Train to persist fitted preprocessors with checkpoints.
```

Comment:
This should be a private API entirely right? Only used by Ray Train?

We don't expect library developers to be using this?

Reply:
That's right, though it should still be DeveloperAPI for cross-library usages.